server/static/dashboard/collab.js

```diff
@@ -7,42 +10,486 @@ collab.fetch(function(err) {
   if (err) { throw err; }
-  
+
+  // Visual 1: total diff
+  // Visual 2: Regex matching
+  // Visual 3: total diff + regex matching
+  // Visual 4: total diff + hide common prefix
+  // Visual 5: total diff + regex matching + hide common prefix
+
+  if (visual[0] == '1' ||
+      visual[0] == '3' ||
+      visual[0] == '4' ||
+      visual[0] == '5') {
+    parameters["threshold"] = getThresholdFromUrl(visual);
+  }  
+
+  if (visual[0] == '2' ||
+      visual[0] == '3' ||
+      visual[0] == '5') {
+    var regexes = getRegexesFromUrl(visual);
+    parameters['regexes'] = regexes;
+    regexes.forEach(function(regex) {
+      addRegexToControls(regex);
+    });
+  }
+
+  if (visual[0] == '4' ||
+      visual[0] == '5') {
+    parameters['hideCommonPrefixes'] = true;
+  }
+
+  showFiles(files, parameters);
+});
+
+/**
+ * Display the files using the provided updateFunction and passing
+ *   extraArgs to that updateFunction.
+ */
+function showFiles(files, extraArgs) {
   var list = document.querySelector('#files');
   files.sort(function(a, b) { return a.data.filepath.localeCompare(b.data.filepath); });
   files.forEach(function(file) {
     var item = document.importNode(document.querySelector('#file').content, true);
+    
     var heading = item.querySelector('h4');
     heading.textContent = file.data.filepath;
     var diff = item.querySelector('.diff code');
     list.appendChild(item);
-    
+
     $.ajax('/baseline/' + project + '/' + file.data.filepath).done(function(baseline) {
+      // Save data for regex updating
+      // TODO: Better way to do this?
+      $(diff).data('baseline', baseline);
+      $(diff).data('text', file.data.text); // TODO: update this data on update
+      $(diff).data('filepath', file.data.filepath);
+
+      var extraArgsForFile = Object.assign({'filepath': file.data.filepath}, extraArgs);
+
       if (cutoff) {
         $.ajax('/historical/' + project + '/' + collabid + '/' + file.data.filepath + '/' + cutoff).done(function(historical) {
-          updateDiff(diff, baseline, historical.data ? historical.data.text : undefined);
+
+          // Used for regex updating
+          if (historical.data) {
+            $(diff).data('text', historical.data.text); // TODO: update this data on update
+          }
+
+          updateFunction(diff, baseline, historical.data ? historical.data.text : undefined, extraArgsForFile);
+
         }).fail(function(req, status, err) {
           diff.textContent = 'Error fetching code: ' + errorToString(req.responseJSON, status, err);
         });
       } else {
         file.subscribe(function() {
-          updateDiff(diff, baseline, file.data.text);
-          file.on('op', function() {
-            updateDiff(diff, baseline, file.data.text);
+          updateFunction(diff, baseline, file.data.text, extraArgsForFile);
+          file.on('op', function(op) {
+            extraArgs["op"] = op;
```

Comment:
This is a bug. Again: should last op be an explicit argument to this `updateFunction`? Which by the way needs to be renamed to something better.

Another possibility is that if `op` is here for auto-scrolling, then auto-scrolling does its own subscription to `file.on('op', ...)`, although that doesn't work without changes to the DOM generated by diffing.

Reply:
Renamed to `displayFileVisual`.